relay/streams.py

```diff
@@ -12,21 +12,63 @@
 
 
 class Client(object):
+    """Represents the connection to a client. Different subscriptions can be connected to the same client"""
 
-    def send(self, id: str, event: Publishable):
+    def __init__(self) -> None:
+        self.subscriptions = []  # type: List[Subscription]
+
+    def register(self, subscription: 'Subscription') -> None:
+        """
+        Registers a subscription that this client has done.
+        On closing the connection with `close` these subscription will get unsubscribed
+        """
+        self.subscriptions.append(subscription)
+
+    def unregister(self, subscription: 'Subscription') -> None:
+        """
+        Unregisters a subscription that this client is not listing for anymore.
+        On closing the connection with `close` these subscription will get unsubscribed
+        """
+        self.subscriptions.remove(subscription)
+
+    def send(self, subscription_id: str, event: Publishable) -> None:
+        """
+        Sends an event to the client that belongs to a subscription with the given subscription id
+        Raises:
+            DisconnectedError: This is raised if the client has already disconnected.
+        """
         raise NotImplementedError
 
+    def close(self) -> None:
+        """
+        Should be called when the connection to this client is closed.
+        This will unsubscribe all registered subscriptions
+        """
+        for subscription in self.subscriptions:
```

Comment:
This looks like we may have to iterate over a copy of self.subscriptions. Could that be the case?

Reply:
I dont think so, as we dont delete from that list while iterating. I can still do it to be sure. 
But I think i should probably delete the subscriptions at the end, which I dont do.

Reply:
Subscription.unsubscribe calls client.unregister, unregister seems to call self.subscriptions.remove(subscription), so this looks like it is modifying that list. Could also be the case that it's another list. I'm not sure either, what is being modified here.

Reply:
Those are two different lists. A subjects keeps track of its subscription, and a client keeps track where he subscribed to

Reply:
I will copy it to be sure.

Reply:
No wait, actually you are right. My bad.